Sources/PackageLoading/PackageBuilder.swift

```diff
@@ -433,6 +440,9 @@ public struct PackageBuilder {
                     if dependency.type != .library {
                         throw ModuleError.executableAsDependency(module: module.name, dependency: name)
                     }
+                    guard module.name != name else {
+                        throw ModuleError.cycleDetected("the target \(name) can't have a dependency on itself.")
+                    }
```

Comment:
We should, it's the more basic problem and will catch the same issues. We can add a new GraphAlgorithms method to find cycles in a graph (happens as a side effect of some of the existing ones, but we might want a standalone version if we want to detect it explicitly here).